crates/stages/src/stages/headers.rs

```diff
@@ -218,6 +221,61 @@ where
 
+        let tip_block_number = match tip {
+            // If tip is hash and it equals to the first downloaded header's hash, we can use
+            // the block number of this header as tip.
+            BlockHashOrNumber::Hash(hash) => downloaded_headers.first().and_then(|header| {
+                if header.hash == hash {
+                    Some(header.number)
+                } else {
+                    None
+                }
+            }),
+            // If tip is number, we can just grab it and not resolve using downloaded headers.
+            BlockHashOrNumber::Number(number) => Some(number),
+        };
+
+        // Since we're syncing headers in batches, gap tip will move in reverse direction towards
+        // our local head with every iteration. To get the actual target block number we're
+        // syncing towards, we need to take into account already synced headers from the database.
+        // It is `None`, if tip didn't change and we're still downloading headers for previously
+        // calculated gap.
+        let target_block_number = if let Some(tip_block_number) = tip_block_number {
+            let local_max_block_number = tx
+                .cursor_read::<tables::CanonicalHeaders>()?
+                .last()?
+                .map(|(canonical_block, _)| canonical_block);
+
+            Some(tip_block_number.max(local_max_block_number.unwrap_or(tip_block_number)))
+        } else {
+            None
+        };
+
+        let mut stage_checkpoint = current_progress
+            .headers_stage_checkpoint()
+            .unwrap_or(HeadersCheckpoint {
+            // If for some reason (e.g. due to DB migration) we don't have `downloaded_headers`
+            // in the middle of headers sync, set it to the local head block number +
+            // number of block already filled in the gap.
+            downloaded_headers: local_head +
+                (target_block_number.unwrap_or_default() - tip_block_number.unwrap_or_default()),
+            // Shouldn't fail because on the first iteration, we download the header for missing
+            // tip, and use its block number.
+            total_headers: target_block_number.unwrap_or_else(|| {
+                warn!(target: "sync::stages::headers", ?tip, "No downloaded header for tip found");
+                // Safe, because `Display` impl for `StageCheckpoint` will fallback to displaying
+                // just `downloaded_headers`
+                0
+            }),
+        });
+
+        // Total headers can be updated if we received new tip from the network, and need to fill
+        // the local gap.
+        if let Some(target_block_number) = target_block_number {
+            stage_checkpoint.total_headers = target_block_number;
+        }
+        stage_checkpoint.downloaded_headers += downloaded_headers.len() as u64;
+
         // Write the headers to db
```

Comment:
we need to populate the stage_checkpoint before writing the headers because of the target block?

Reply:
yes, but also `downloaded_headers` is consumed on headers write